src/Bicep.LangServer/Completions/PublicRegistryModuleMetadataProvider.cs

```diff
@@ -16,69 +17,124 @@ namespace Bicep.LanguageServer.Providers
                 {
-                    moduleMetadataCache = metadata;
+                    return;
                 }
+
+                this.isQueryingLiveData = true;
             }
-            catch (Exception e)
+
+            var _ = Task.Run(async () =>
             {
-                Trace.WriteLine(string.Format("Encountered following exception while intializing MCR modules metadata: {0}", e.Message));
-            }
+                try
+                {
+                    if (initialDelay)
+                    {
+                        await Task.Delay(5000);
+                    }
 
-            Trace.WriteLine("Unable to initialize MCR modules metadata.");
-        }
+                    var modules = await TryGetModulesLive();
 
-        public async Task<IEnumerable<string>> GetModuleNames()
-        {
-            List<string> moduleNames = new List<string>();
+                    this.cachedModules = modules;
+                    this.lastSuccessfulQuery = DateTime.Now;
+                }
+                finally
+                {
+                    // Throttle unsuccessful requests to avoid spamming the endpoint
+                    await Task.Delay(5000);
 
-            if (!moduleMetadataCache.Any())
-            {
-                await Initialize();
-            }
+                    lock (this.queryingLiveSyncObject)
+                    {
+                        Trace.Assert(this.isQueryingLiveData, "this.isQueryingLiveData should be true");
+                        this.isQueryingLiveData = false;
+                    }
+                }
+            });
+        }
 
-            foreach (var moduleMetadataCacheEntry in moduleMetadataCache)
+        private bool IsCacheExpired()
+        {
+            var expired = this.lastSuccessfulQuery.HasValue && this.lastSuccessfulQuery.Value + this.cacheValidFor < DateTime.Now;
+            if (expired)
             {
-                moduleNames.Add(moduleMetadataCacheEntry.moduleName);
+                Trace.TraceInformation("Public modules cache is expired.");
             }
 
-            return moduleNames;
+            return expired;
         }
 
-        public async Task<IEnumerable<string>> GetVersions(string moduleName)
+        private async Task<ModuleMetadata[]> TryGetModulesLive()
         {
-            List<string> versions = new List<string>();
+            Trace.WriteLine($"Retrieving list of public registry modules...");
 
-            if (!moduleMetadataCache.Any())
+            try
             {
-                await Initialize();
-            }
-
-            ModuleMetadata? metadata = moduleMetadataCache.FirstOrDefault(x => x.moduleName.Equals(moduleName, StringComparison.Ordinal));
+                HttpClient httpClient = new HttpClient();
+                var moduleMetadata = await httpClient.GetStringAsync(LiveDataEndpoint);
+                var metadata = JsonConvert.DeserializeObject<ModuleMetadata[]>(moduleMetadata);
```

Comment:
We've been trying to use System.Text.Json instead of Newtonsoft for new code. Would also be good to use a stream rather than a string:

```csharp
using var metadataStream = await httpClient.GetStreamAsync(LiveDataEndpoint);
var metadata = JsonSerializer.Deserialize<ModuleMetadata[]>(metadataStream);
```